# Prédiction de la valeur marchande d'un joueur

Ce notebook permet d'estimer la valeur marchande d'un joueur de football à partir du modèle final
entraîné dans le notebook `09_modelisation_finale.ipynb`.

**Il est autosuffisant** : il ne dépend que d'un seul fichier, `../modelisation/pipeline_explicabilite.pkl`
(généré à la fin du notebook 9), qui contient les modèles déjà entraînés. Pas besoin de ré-entraîner quoi
que ce soit ni de relancer les autres notebooks.

## Partie A : Chargement du modèle et des données

On recharge tout ce qui est nécessaire depuis le pipeline sauvegardé par le notebook 9. Rien d'autre
n'est requis (pas de ré-entraînement, pas de fichier de configuration supplémentaire).

In [ ]:
import numpy as np
import pandas as pd
import difflib
import os
import sys
from datetime import datetime

import joblib
from sklearn.base import BaseEstimator, RegressorMixin

# On connecte le notebook à tous les fichiers inclus dans le dossier /fonctions
sys.path.insert(0, os.path.abspath("../fonctions/modelisations"))

%load_ext autoreload
%autoreload 2

# On importe les fonctions utiles à la modélisation
from prediction_joueur import *

pipeline = joblib.load("../modelisation/pipeline_explicabilite.pkl")

X_train = pipeline["X_train"]
X_val = pipeline["X_val"]
X_test = pipeline["X_test"]
X_en_cours = pipeline["X_en_cours"]

df_train = pipeline["df_train"]
df_val = pipeline["df_val"]
df_test = pipeline["df_test"]
df_en_cours = pipeline["df_en_cours"]

modeles_finaux = pipeline["modeles_finaux"]
meta = pipeline["meta"]
colonne_cible = pipeline["colonne_cible"]

colonne_joueur = "player"
colonne_saison = "season_year"
colonne_team = "team"

print("Pipeline chargé avec succès.")
print(f"{X_train.shape[1]} variables attendues par le modèle.")

Pipeline chargé avec succès.
146 variables attendues par le modèle.


## Partie B : Le modèle de prédiction (Stacking)

Le modèle final combine 3 modèles (XGBoost, LightGBM, CatBoost) : chacun propose une estimation, puis un méta-modèle apprend la meilleure combinaison des trois pour donner la prédiction finale, directement en euros.

In [3]:
modele_stack = StackingModel(
    models=[
        modeles_finaux["XGBoost (log)"],
        modeles_finaux["LightGBM (log)"],
        modeles_finaux["CatBoost (log)"],
    ],
    meta_model=meta,
)

print("Modèle de stacking prêt.")

Modèle de stacking prêt.


## Partie C : Manipulations du modèle

### C.1 : Rechercher un joueur existant

Recherche un joueur par son nom parmi toutes lesnsaisons disponibles (train / validation / test / en cours). Par défaut, la saison la plus récente disponible est utilisée.

In [4]:
# Historique de toutes les prédictions réalisées dans cette session
historique_predictions = []

DATASETS_PAR_SPLIT = {
    "train": (df_train, X_train),
    "val": (df_val, X_val),
    "test": (df_test, X_test),
    "en_cours": (df_en_cours, X_en_cours),
}

**Exemple d'utilisation** :

In [ ]:
predire_joueur_existant("Kylian Mbappé", DATASETS_PAR_SPLIT, modele_stack, historique_predictions)

# Pour une saison précise (au lieu de la plus récente) :
predire_joueur_existant("Kylian Mbappé", DATASETS_PAR_SPLIT, modele_stack, historique_predictions, saison=2022)

6 saisons trouvées pour Kylian Mbappé, on utilise la plus récente (2025). Précisez saison=... pour en choisir une autre.
 Kylian Mbappé
   Saison               : 2025
    Valeur marchande prédite : 111 381 703 €
    Valeur réelle connue    : 200 000 000 €
    Écart                   : -44.3 %
 Kylian Mbappé
   Saison               : 2022
    Valeur marchande prédite : 148 094 625 €
    Valeur réelle connue    : 180 000 000 €
    Écart                   : -17.7 %


np.float64(148094625.2091464)

### C.2 : Saisir un joueur manuellement

Pour simuler un joueur (nouveau, hypothétique, ou dont on ne connaît que quelques caractéristiques), il n'est **pas nécessaire de renseigner les 146 variables du modèle**. On part d'un profil "joueur moyen" (médiane des variables numériques, valeur la plus fréquente pour les variables 0/1, calculées automatiquement sur les données d'entraînement), et on ne renseigne que ce que l'on sait.

**Champs disponibles pour la saisie**

| Clé à utiliser dans le dictionnaire | Description | Exemple de valeur |
|---|---|---|
| `age` | Âge du joueur | `24` |
| `taille_cm` | Taille en cm | `182` |
| `pied` | `"Droit"`, `"Gauche"` ou `"Ambidextre"` | `"Gauche"` |
| `poste` | `"Gardien"`, `"Défenseur"`, `"Milieu"` ou `"Attaquant"` | `"Attaquant"` |
| `championnat` | `"Premier League"`, `"La Liga"`, `"Ligue 1"`, `"Bundesliga"`, `"Serie A"` ou `"Autre / non renseigné"` | `"Ligue 1"` |
| `valeur_marchande_precedente_euros` | Valeur marchande de la saison précédente (facteur le plus important du modèle) | `18_000_000` |
| `classement_equipe` | Classement de l'équipe en championnat | `5` |
| `buts` | Buts marqués dans la saison | `12` |
| `passes_decisives` | Passes décisives dans la saison | `6` |
| `matchs_joues` | Nombre de matchs disputés | `30` |
| `minutes_jouees_par_match` | Minutes jouées par match disputé | `75` |

Tout champ non renseigné garde sa valeur par défaut (profil moyen).

**Exemple d'utilisation** :

In [6]:
saisie_exemple = {
    "age": 20,
    "taille_cm": 180,
    "pied": "Gauche",
    "poste": "Attaquant",
    "championnat": "Ligue 1",
    "valeur_marchande_precedente_euros": 12_000_000,
    "classement_equipe": 4,
    "buts": 15,
    "passes_decisives": 5,
    "matchs_joues": 32,
    "minutes_jouees_par_match": 78,
}

predire_joueur_manuel(modele_stack, historique_predictions, saisie_exemple, X_train, nom_affiche="Exemple - jeune attaquant prometteur")

 Exemple - jeune attaquant prometteur
    Valeur marchande prédite : 11 053 285 €


np.float64(11053284.868388165)

### C3 : Export des prédictions de cette session

Toutes les prédictions réalisées ci-dessus (recherche ou saisie manuelle) sont exportées ici.

In [7]:
os.makedirs("../exports", exist_ok=True)

df_export = pd.DataFrame(historique_predictions)
chemin_csv = "../exports/predictions_session_powerbi.csv"

df_export.to_csv(chemin_csv, index=False, encoding="utf-8-sig", decimal=",")

print(f"{len(df_export)} prédiction(s) de cette session exportée(s) vers {chemin_csv}")
df_export

3 prédiction(s) de cette session exportée(s) vers ../exports/predictions_session_powerbi.csv


,date_prediction,joueur,mode,saison,valeur_predite_euros,valeur_reelle_euros,ecart_pct
0,2026-09-22 09:34,Kylian Mbappé,recherche,2025.0,1.113817e+08,200000000.0,-44.31
1,2026-09-22 09:34,Kylian Mbappé,recherche,2022.0,1.480946e+08,180000000.0,-17.73
2,2026-09-22 09:34,Exemple - jeune attaquant prometteur,manuel,NaN,1.105328e+07,NaN,NaN


### C.4 : Export complet de tous les joueurs

Pour visualiser dans PowerBI l'évolution des VM de **tous** les joueurs (pas seulement ceux testés manuellement plus haut), on calcule ici la VM réelle **et** la VM prédite par le modèle pour chaque ligne joueur/saison disponible (train + validation + test + en cours). Cela permet, dans PowerBI, de comparer visuellement valeur réelle et valeur estimée par le modèle, saison après saison.

In [8]:
df_historique = construire_export_historique(datasets = DATASETS_PAR_SPLIT, modele_stack = modele_stack)

chemin_historique = "../exports/historique_vm_joueurs_powerbi.csv"
df_historique.to_csv(chemin_historique, index=False, encoding="utf-8-sig", decimal=",")

print(f"{len(df_historique)} lignes joueur/saison exportées vers {chemin_historique}")
df_historique.head()

14413 lignes joueur/saison exportées vers ../exports/historique_vm_joueurs_powerbi.csv


,joueur,equipe,saison,jeu,vm_reelle_euros,vm_predite_euros,ecart_pct
0,Aaron Connolly,Brighton,2020,train,7000000.0,9.597575e+06,0.37
1,Aaron Cresswell,West Ham United,2020,train,5000000.0,8.837039e+06,0.77
2,Aaron Cresswell,West Ham United,2021,train,3000000.0,4.256691e+06,0.42
3,Aaron Cresswell,West Ham United,2022,train,1200000.0,1.859468e+06,0.55
4,Aaron Hickey,Bologna,2020,train,5000000.0,5.208645e+06,0.04
